### PR TITLE
int4: disable const_folding for unpack_int4

### PR DESCRIPTION
**Pull request by lakhinderwalia**
disable const_folding for unpack_int4

### CODE REVIEW

File: src/quantize_int4.cpp
Comment (umangyadav) [inline src/quantize_int4.cpp:58]: tidy will complain on `||`

File: src/driver/main.cpp
Comment (umangyadav) [inline src/driver/main.cpp:506]: Better to name it as `--int4-weights`

File: src/quantize_int4.cpp
Comment (lakhinderwalia) [reply src/quantize_int4.cpp:58]: Yes. of course! Thanks.

Comment (pfultz2) [reply src/quantize_int4.cpp:58]: There is no need to check for scalar. If its scalar then it is broadcasted.

File: src/propagate_constant.cpp
Comment (umangyadav) [inline src/propagate_constant.cpp:48]: Would this work if let's say there is a "reshape" or some other operation in between `unpack_int4` and `dequantizelinear` ? 
e.g. `unpack_int4 --> convert --> DQ`

File: src/driver/main.cpp
Comment (lakhinderwalia) [reply src/driver/main.cpp:506]: Would you consider dot operations to contain 'weights'?

Comment (umangyadav) [reply src/driver/main.cpp:506]: Yes

File: src/propagate_constant.cpp
Comment (pfultz2) [reply src/propagate_constant.cpp:48]: Why would there be a convert?

Comment (umangyadav) [reply src/propagate_constant.cpp:48]: Contrived case i came up with.

Need to handle general case if possible pre-emptively.

Comment (umangyadav) [reply src/propagate_constant.cpp:48]: It can any other op. "convert' is just an example

Comment (pfultz2) [reply src/propagate_constant.cpp:48]: We are the ones that will insert these operators either in quantize_int4 or in the onnx parser, so i dont understand why we would insert a `convert`, since it wouldn't work.

Comment (umangyadav) [reply src/propagate_constant.cpp:48]: model could also come from quantizer tool e.g. vitisAI quantizer.

File: src/quantize_int4.cpp
Comment (pfultz2) [inline src/quantize_int4.cpp:71]: You can do `inp->eval().to_vector<float>()`.

**Comment by lakhinderwalia** [comment]
@pfultz2, @CharlieL7, this PR is being moved out of the draft mode. Some of its necessary functionality is in other PRs. For example, [block_quantization support.](https://github.com/ROCm/AMDMIGraphX/pull/3412)  Also int4 (i.e. [signed support](https://github.com/ROCm/AMDMIGraphX/pull/3359)) for the basic pack/unpack. [Parse int4 graph](https://github.com/ROCm/AMDMIGraphX/pull/3408).
At this stage, we are able to compile the test graphs that are checked-in, and also other full quark-generated graph. We also need some of these other PRs to merge first. Some other tests will be added as the review progresses. Thanks.

File: src/propagate_constant.cpp
Comment (pfultz2) [inline src/propagate_constant.cpp:39]: Dont use a global variable, this is not thread-safe, and this pass is called across multiple threads.

File: src/propagate_constant.cpp
Comment (pfultz2) [inline src/propagate_constant.cpp:135]: Dont refactor propagate_constant, there is already perf regressions across almost all models because its skipping a lot of const folding. 

For now just add dequatizelinear and unpack_int4 to skip_propagate like you did in earlier:

```cpp
bool skip_propagate(instruction_ref ins)
{
    if(ins->name() == "contiguous" or ins->name() == "dequantizelinear")
        return skip_propagate(ins->inputs().front());
    if(ins->name() == "unpack_int4")
        return true;
    auto&& s = ins->get_shape();
    if(s.broadcasted() and not s.scalar())
        return true;
    if(s.scalar() and s.elements() != 1)
        return true;
    return false;
}
```

This will handle the cases with `--int4-weights`. It wont handle other cases from other models, but we can tackle that in a later PR(I dont want this to block the other changes that are being made in this PR) as we need to discuss a better design on handling those cases including the additional cases for block quantization.

File: src/simplify_qdq.cpp
Comment (pfultz2) [inline src/simplify_qdq.cpp:445]: Add unit tests for this pass.

File: src/targets/gpu/mlir.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir.cpp:650]: This shouldn't override the axis value, but we really dont use this.

Ideally, we should just unpack across the fastest axis which is not always the same as the last axis. Let me open an issue to address this in the future.

File: src/targets/gpu/target.cpp
Comment (pfultz2) [inline src/targets/gpu/target.cpp:141]: This should not be moved up here. This should happen in optimize_module, like always. I dont think this is needed to support `--int4-wieghts`.

File: src/targets/gpu/target.cpp
Comment (pfultz2) [inline src/targets/gpu/target.cpp:136]: I dont really like having just in4 specific comments here as these passes cover more than just int4 cases, so I think it is confusing.

File: src/include/migraphx/quantization.hpp
Comment (pfultz2) [inline src/include/migraphx/quantization.hpp:52]: This should be called `quantize_int4_weights` since it will only quantize the weights.

**Review by pfultz2 — CHANGES_REQUESTED** [review]

File: src/propagate_constant.cpp
Comment (lakhinderwalia) [reply src/propagate_constant.cpp:135]: Thanks. Yes, this simple case was tackled early on. But it was too far removed from what the actual quark-generated model did with its int4 zero_points, which was expected to be supported in the first-deliverable itself.

File: src/targets/gpu/mlir.cpp
Comment (lakhinderwalia) [reply src/targets/gpu/mlir.cpp:650]: Agreed, we don't really use the axis for unpack/pack.

File: src/targets/gpu/target.cpp
Comment (lakhinderwalia) [reply src/targets/gpu/target.cpp:141]: (Yes, this mayn't be necessary for supporting just the `--int4-weights`.)
This was moved up due to e.g. this simple test case needs to be supported: `test/onnx/int4_const_identity_block_sz_2_qdq_test.onnx` (new file in this PR).

File: src/propagate_constant.cpp
Comment (lakhinderwalia) [reply src/propagate_constant.cpp]: I will mark this resolved for now. Don't see any reshape after `unpack_int4`. Thanks.

File: src/targets/gpu/target.cpp
Comment (lakhinderwalia) [reply src/targets/gpu/target.cpp:136]: Reverted.

File: src/simplify_qdq.cpp
Comment (lakhinderwalia) [reply src/simplify_qdq.cpp:445]: Added.

File: src/propagate_constant.cpp
Comment (lakhinderwalia) [reply src/propagate_constant.cpp]: For now, the changes are reverted to handing just `--int4-weights`.

File: src/targets/gpu/target.cpp
Comment (pfultz2) [inline src/targets/gpu/target.cpp:170]: I think this should be removed, its not necessary for int4-weights, and I am concerned it might cause perf regressions.

File: test/gpu/mlir.cpp
Comment (pfultz2) [inline test/gpu/mlir.cpp:487]: This isn't very readable. Format this string better like the other tests do.

File: src/quantize_int4.cpp
Comment (pfultz2) [reply src/quantize_int4.cpp:71]: Are you able to update this as well? @lakhinderwalia

File: src/targets/gpu/target.cpp
Comment (lakhinderwalia) [reply src/targets/gpu/target.cpp:165]: You are right. I have missed pushing this particular file! Let us wait what the perf results are anyways!

File: src/quantize_int4.cpp
Comment (lakhinderwalia) [reply src/quantize_int4.cpp:71]: I haven't done this yet. That is why the comment is left open. Will add it to my next set of changes. Thanks.

File: test/gpu/mlir.cpp
Comment (lakhinderwalia) [reply test/gpu/mlir.cpp:487]: Yes. Will do. We did talk about it.

File: src/targets/gpu/target.cpp
Comment (lakhinderwalia) [reply src/targets/gpu/target.cpp]: It has been removed. However, I didn't notice any perf regression below.

File: src/quantize_int4.cpp
Comment (lakhinderwalia) [reply src/quantize_int4.cpp]: Done.

Comment (lakhinderwalia) [reply src/quantize_int4.cpp:71]: This change gave a compile failure. I just reverted it for now.

**Review by pfultz2 — APPROVED** [review]
This is still missing unit test for simplify_qdq. It really should be addressed either before merging or after merging.

**Comment by lakhinderwalia** [comment]
> This is still missing unit test for simplify_qdq. It really should be addressed either before merging or after merging.

Added.

File: test/int4_test.cpp
Comment (pfultz2) [inline test/int4_test.cpp:113]: This should be moved with the other simplify_qdq tests, but lets do that later. I think this should just be merged in now.

**Comment by pfultz2** [comment]
There is an mlir failure on jenkins:

```
[2024-09-19T21:19:12.937Z] [   RUN    ] int4_unpack_conv

[2024-09-19T21:19:12.937Z] terminate called after throwing an instance of 'migraphx::version_1::exception'

[2024-09-19T21:19:12.937Z]   what():  /home/jenkins/workspace/AMDMIGraphX_PR-3322/src/targets/gpu/mlir.cpp:787: run_backend_pipeline: MLIR backend compilation failed: Error: The size of rock.alloc should be greather than zero.

[2024-09-19T21:19:12.937Z] Note: see current operation: %43 = "rock.alloc"() : () -> memref<16xi4, #gpu.address_space<private>>
```

Do you see this same error locally?

**Comment by lakhinderwalia** [comment]
> There is an mlir failure on jenkins:
> 
> ```
> [2024-09-19T21:19:12.937Z] [   RUN    ] int4_unpack_conv
> 
> [2024-09-19T21:19:12.937Z] terminate called after throwing an instance of 'migraphx::version_1::exception'
> 
> [2024-09-19T21:19:12.937Z]   what():  /home/jenkins/workspace/AMDMIGraphX_PR-3322/src/targets/gpu/mlir.cpp:787: run_backend_pipeline: MLIR backend compilation failed: Error: The size of rock.alloc should be greather than zero.
> 
> [2024-09-19T21:19:12.937Z] Note: see current operation: %43 = "rock.alloc"() : () -> memref<16xi4, #gpu.address_space<private>>
> ```
> 
> Do you see this same error locally?

No.

**Comment by pfultz2** [comment]
If this needs a change from mlir to work, then just comment out the broken test for now so we can merge this in.

**Comment by lakhinderwalia** [comment]
Thanks. Actually, MLIR has a fix just commited-in. But I can disable the test. However, we do need one more approval :-)

**Comment by causten** [comment]
I've got https://github.com/ROCm/AMDMIGraphX/pull/3467 which is pulling from mlir tip.  I"m planning to merge it asap

**Comment by lakhinderwalia** [comment]
> I've got #3467 which is pulling from mlir tip. I"m planning to merge it asap

@causten, please include 1b99123a4e86629b07b1c2815668f6222d093d70 or newer. Thanks.

**Review by shivadbhavsar — APPROVED** [review]

**Pull request merged**
